File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/rendering/A2UIRendererView.kt
package com.interfaceai.a2ui.rendering

import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import com.interfaceai.a2ui.models.Surface
import com.interfaceai.a2ui.processing.DataBindingResolver
import com.interfaceai.a2ui.processing.MessageProcessor
import com.interfaceai.a2ui.processing.TokenResolver
import kotlinx.serialization.json.JsonElement

/**
 * Top-level Composable that renders an A2UI surface.
 *
 * Usage:
 * ```kotlin
 * val processor = MessageProcessor()
 * val surface = processor.processJSON(jsonString)
 * if (surface != null) {
 *     A2UIRendererView(surface = surface)
 * }
 * ```
 */
@Composable
fun A2UIRendererView(
    surface: Surface,
    onAction: ((String, Map<String, JsonElement>?) -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val tokenResolver = TokenResolver(surface.designTokens)
    val dataResolver = DataBindingResolver(surface.dataModel)

    ComponentNodeView(
        componentId = surface.rootComponentId,
        surface = surface,
        tokenResolver = tokenResolver,
        dataResolver = dataResolver,
        onAction = onAction,
        modifier = modifier
    )
}

/**
 * Convenience: render from raw JSON string.
 */
@Composable
fun A2UIFromJSON(
    json: String,
    onAction: ((String, Map<String, JsonElement>?) -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    var surface by remember { mutableStateOf<Surface?>(null) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(json) {
        try {
            val processor = MessageProcessor()
            surface = processor.processJSON(json)
        } catch (e: Exception) {
            error = e.message
        }
    }

    when {
        surface != null -> A2UIRendererView(
            surface = surface!!,
            onAction = onAction,
            modifier = modifier
        )
        error != null -> Text("Error: $error", color = Color.Red)
        else -> CircularProgressIndicator()
    }
}

File: ios-renderer/Sources/A2UIRenderer/Rendering/A2UIRendererView.swift
import SwiftUI

/// Top-level SwiftUI view that renders an A2UI surface.
///
/// Usage:
/// ```swift
/// let processor = MessageProcessor()
/// try processor.processJSON(jsonString)
/// if let surface = processor.surfaces["widget_hello"] {
///     A2UIRendererView(surface: surface)
/// }
/// ```
public struct A2UIRendererView: View {
    let surface: Surface
    var onAction: ((String, [String: JSONValue]?) -> Void)?

    public init(surface: Surface, onAction: ((String, [String: JSONValue]?) -> Void)? = nil) {
        self.surface = surface
        self.onAction = onAction
    }

    public var body: some View {
        let tokenResolver = TokenResolver(tokens: surface.designTokens)
        let dataResolver = DataBindingResolver(dataModel: surface.dataModel)

        ComponentNodeView(
            componentId: surface.rootComponentId,
            surface: surface,
            tokenResolver: tokenResolver,
            dataResolver: dataResolver,
            onAction: onAction
        )
    }
}

/// Convenience: render from raw JSON data.
public struct A2UIFromJSON: View {
    @State private var processor = MessageProcessor()
    @State private var surfaceId: String?
    @State private var error: String?
    let json: Data
    var onAction: ((String, [String: JSONValue]?) -> Void)?

    public init(json: Data, onAction: ((String, [String: JSONValue]?) -> Void)? = nil) {
        self.json = json
        self.onAction = onAction
    }

    public var body: some View {
        Group {
            if let surfaceId, let surface = processor.surfaces[surfaceId] {
                A2UIRendererView(surface: surface, onAction: onAction)
            } else if let error {
                Text("Error: \(error)")
                    .foregroundColor(.red)
                    .font(.caption)
            } else {
                ProgressView()
            }
        }
        .task {
            do {
                let surface = try processor.processJSON(json)
                surfaceId = surface?.surfaceId
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
}
